modules/gapi/include/opencv2/gapi/own/assert.hpp

```diff
@@ -33,7 +40,7 @@ namespace detail
 
 
 #ifdef NDEBUG
-#  define GAPI_DbgAssert(expr)
+#  define GAPI_DbgAssert(expr) cv::util::suppress_unused_warning(expr)
```

Comment:
AFAIK, expression is still should be computed by compiler (and then dead code is eliminated - but not all).

Reply:
As `suppress_unused_warning(expr)` is the function call, it's actual argument i.e. `expr`  _is_ calculated, then however  it's value is discarded as bound to unnamed function argument. 

The `suppress_unused_warning` defined as:

```c++
template<typename T> void suppress_unused_warning( const T& ) {}
```

Reply:
All functions and methods are not "pure" in C++ standard ("pure" is GCC extension).
So some non-inline `.type()` calls will be evaluated if they are part of this expr.

Reply:
Please follow the classic C `assert()` implementation from `assert.h` / `cassert` headers (it is "dbg" assert in OpenCV):
https://en.cppreference.com/w/cpp/error/assert

Reply:
Having the assert implemented like this does not prevent "unused" warnings (and that is why it was not enough to use `CV_DbgAssert` as is and I had to re-invent it. )

Reply:
it seems that `sizeof()` is the solution here. As it is silence the warning and does not evaluate it's argument

Reply:
> it seems that sizeof() is the solution here. 

But not for lambdas :)  - fixed to use `constexpr` and bool operations  short-circuit evaluation